### DISCUSSION TITLE
PKCS12_create API in openssl 30 compatibility with Java 8 API

### DISCUSSION THREAD

**Discussion by ChauhanPriyanka** [openssl/openssl#21942] [Q&A] [answered]
Hi team, our application is providing certificates to android devices.
Currently we are upgrading the c++ application to openssl30. In order to provide backward compatibility we have loaded default and legacy provider in our application. Certificates are generated however android devices failed to parse the certificate reason being the Java8 keystore apis are not compatible with openssl 30.
https://stackoverflow.com/questions/70244066/keytool-error-java-io-ioexception-parsealgparameters-failed-objectidentifier.

We are already loading legacy provider and default provider to generate certificate, could you please help shall we disable default provider & use only legacy provider before PKCS12_create operation 

```
OpenSSL1.0.2
MAC: sha1, Iteration 1
MAC length: 20, salt length: 8
PKCS7 Encrypted data: pbeWithSHA1And40BitRC2-CBC, Iteration 2048
```

```
OpenSSL3.0

MAC: sha256, Iteration 2048
MAC length: 32, salt length: 8
PKCS7 Encrypted data: PBES2, PBKDF2, AES-256-CBC, Iteration 2048, PRF hmacWithSHA256
```

**Comment by ChauhanPriyanka** [comment]
@t8m @paulidale @mattcaswell, could you please help here

**Comment by ChauhanPriyanka** [comment]
When default and legacy provider both are loaded in a windows c++ application. 
PKCS12_create API will use default provider or legacy provider. 
Please help we are blocked here.

**Comment by mattcaswell** [comment]
Try setting the `nid_key` and `nid_cert` parameters in the call to `PKCS12_create` to the value `NID_pbe_WithSHA1And40BitRC2_CBC`.

See:
https://www.openssl.org/docs/man3.1/man3/PKCS12_create.html

**Comment by ChauhanPriyanka** [comment]
HI @mattcaswell,

Thank you for the response.

I have modified the current code like this -
`PKCS12_create((pswd, szCertStorePath, pkey, cert, NULL, **NID_pbe_WithSHA1And40BitRC2_CBC**, **NID_pbe_WithSHA1And40BitRC2_CBC**,0,0,0);`

Cert was generated and I got following response while reading the pfx file.

```
**openssl pkcs12 -info -in cert.pfx -noout**
MAC: sha256, Iteration 2048
MAC length: 32, salt length: 8
PKCS7 Encrypted data: pbeWithSHA1And40BitRC2-CBC, Iteration 2048
Error outputting keys and certificates
046C0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto\evp\evp_fetch.c:373:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()
```

For clarity, can we say that in OpenSSL102, PKCS12_create uses pbeWithSHA1And40BitRC2 as default. 
Apologies for asking the silly question, I have very limited knowledge of openssl. 

The existing code calls the API like this NID_key & NID_cert were set to 0.
p12 = PKCS12_create(pswd, szCertStorePath, pkey, cert, NULL, 0,0,0,0,0);

This is the generic code, I'm not sure what will be the impact of explicitly setting nid_key & nid_cert to value NID_pbe_WithSHA1And40BitRC2_CBC. It could start working for android device. However the certificate generated from here are consumed by IOS, android and windows devices.

**Reply by mattcaswell** [reply]
> For clarity, can we say that in OpenSSL102, PKCS12_create uses pbeWithSHA1And40BitRC2 as default.

Yes, that is the default in older OpenSSL versions. It's a very weak insecure algorithm which is why the default was changed. However it should work for most applications.

> 046C0000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto\evp\evp_fetch.c:373:Global default library context, Algorithm (RC2-40-CBC : 0), Properties ()

This means that it cannot find the RC2-40-CBC algorithm. That will be because the legacy provider is not loaded. Add the `-legacy` option to your command line.

**Reply by t8m** [reply]
You can try NID_pbe_WithSHA1And3_Key_TripleDES_CBC instead.

**Reply by t8m** [reply]
That should still be very much compatible.

**Comment by ChauhanPriyanka** [comment]
Thank you @mattcaswell.
It is working after adding legacy provider in command line
```
 openssl pkcs12 -info -in cert.pfx -noout -legacy
MAC: sha256, Iteration 2048
MAC length: 32, salt length: 8
PKCS7 Encrypted data: pbeWithSHA1And40BitRC2-CBC, Iteration 2048
Certificate bag
PKCS7 Data
Shrouded Keybag: pbeWithSHA1And40BitRC2-CBC, Iteration 2048
```
Hi @t8m, I will try this NID_pbe_WithSHA1And3_Key_TripleDES_CBC.

The only concern is - I'm modifying generic code and forcing to use specific algo. In future this could have other implications.
p12 = PKCS12_create(pswd, szCertStorePath, pkey, cert, NULL, 0,0,0,0,0); 

Could you help me to find alternate way like disabling/unloading default provider and only use legacy provider for this operation.

**Reply by t8m** [reply, marked as answer]
You have to hard-code the legacy algorithm there. Just loading/unloading the legacy provider does not change the algorithm used by PKCS12_create.

**Comment by ChauhanPriyanka** [comment]
Alright, thank you @t8m for quick response. 
I will try hard-coding of NID_pbe_WithSHA1And40BitRC2_CBC/ NID_pbe_WithSHA1And3_Key_TripleDES_CBC both the above algos and check the compatibility in different devices.

**Comment by ChauhanPriyanka** [comment]
hi @t8m, @mattcaswell,
We did a quick test with the new cert generated with **NID_pbe_WithSHA1And40BitRC2_CBC**
Now in android device we are getting this exception -
` java.io.IOException: Integrity check failed: java.security.NoSuchAlgorithmException: Algorithm HmacPBESHA256`.

Any suggestion here what needs to be changed ?

**Comment by ChauhanPriyanka** [comment]
Used the following code to change to SHA1 for backward compatibility.
1 == PKCS12_set_mac(p12, pswd, strlen(pswd), NULL, 0, 1, EVP_sha1()) & this fixed the above exception
 java.io.IOException: Integrity check failed: java.security.NoSuchAlgorithmException: Algorithm HmacPBESHA256.
